### DISCUSSION TITLE
Docker 部署时遇到文件夹权限问题

### DISCUSSION THREAD

**Discussion by weremexii** [y-young/nazurin#58] [Q&A] [answered]
仅配置 .env 文件（不使用本地存储），使用 docker-compose 启动服务，第一次运行时：
1. docker 会在当前目录下创建用户和用户组都为 root 的 data 文件夹
2. bot 在下载图片之前会试图在 data 下创建 temp 文件夹，触发 Permission denied 错误

第 1 个问题似乎是 docker 的预期行为，可以通过事先以当前用户创建 data 文件夹避免

第 2 个问题在以当前用户创建 data 文件夹后仍会发生。参考 `docker-compose top` 打印的进程信息和 Dockerfile 的内容，我觉得问题在于 python 以 uid 为 5678 的用户运行，因此无法在属于当前用户或者 root 的 data 文件夹下进行操作

docker-compose.yml 中可以指定容器运行用户（可以覆盖 Dockerfile 中指定的用户），我找到的解决方案是
+ 事先以当前用户创建 data 文件夹
+ 在 docker-compose.yml 中为 nazurin 服务添加配置项 `user: $(UID_GID)` 
+ 以命令 `UID_GID="$(id -u):$(id -g)" docker-compose up -d` 启动容器

PS：docker-compose.yml 无法使用 $UID，因为 UID 属于 shell 变量而非环境变量；而以 `UID=$UID docker-compose up -d` 启动会触发 bash 的错误 `UID: readonly variable`，参见 [Stackoverflow](https://stackoverflow.com/questions/56844746/how-to-set-uid-and-gid-in-docker-compose)

我想了解的是以这种解决方案配置 docker-compose.yml 是否合适以及是否是因为我对主机的配置有误导致 Permission denied 错误

**Comment by y-young** [comment, marked as answer]
这个其实是已知问题，只是之前没有找到合适的解决方案，所以一直搁置着。在事先创建好data目录且有root权限的情况下应该没有什么大问题，你提到的第二个问题可能是因为other用户组没有data目录的读写权限。

主要的问题在于Docker容器内外用户不一致，如果不事先创建目录的话Docker会自动创建一个属于root用户的目录，由于容器内为了安全考虑特地切换成了非root用户，所以无法访问data目录。如果容器内依旧以root用户运行的话应该就没有第一个问题，但仍有一个问题是主机上的非root用户无法访问Docker创建的data目录，所以事先创建目录应该是一个合理的选择。

至于使用环境变量配置运行用户的方法，之前是考虑到需要加额外参数比较麻烦，所以没有采用，现在来看还算是一个能够接受的方案，待我进一步研究一下，如果还有其他更好的方案也欢迎提出。应当避免的是在构建Docker镜像时指定用户，因为这需要每个用户自行构建镜像。

**Comment by love2wind** [comment]
感谢提供解决思路，我直接新建了data和data/temp目录，然后还是出现了下面的错误
```
Error: [Errno 13] Permission denied: 'data/temp/wallhaven-85xjpk.jpg'
```
我将data/temp目录权限设置成root777问题解决。
